### PR TITLE
Add new APIs for code gen

### PR DESCRIPTION
**Pull request by steveharter**
Per offline reviews, this is a first pass at adding new public APIs to support the code-gen effort.

Primarily this exposes the existing internal metadata classes (JsonPropertyInfo\JsonClassInfo) as public and also extends these existing classes by adding derived classes (for value-based, object-based and collection-based types) along with helper methods to initialize the metadata and (de)serialize.

There is still work remaining including:
- API review \ usability.
- Renaming of existing classes including `ReadStack`, `WriteStack` and `JsonClassInfo`.
- Add additional helper method that will be called by code-gen if there are left-over JSON that was not deserialized.
- Fallback to reflection to obtain metadata for unsupported cases (fields, ctor parameters)
- Support for circular Type references (a lazy-loaded approach for metadata).
- Additional tests: circular references, preserve references, custom converters, benchmarks (especially startup time).
- Design overlap with custom converters and to support the linker for code removal.
  - May add a new constructor to `JsonSerializerOptions` that will not initialize the built-in converters and prevent the Emit- and Reflection-based `MemberAccessor` strategies from being used.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs [hunk 1]
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
+using System.Runtime.CompilerServices;
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:5]: nit: why this using?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectCodeGenConverter.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+            JsonClassInfo jsonClassInfo = state.Current.JsonClassInfo;
+            Debug.Assert(jsonClassInfo.Deserialize != null);
+
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+            }
+
+            object objValue = jsonClassInfo.Deserialize!(ref reader, ref state, jsonClassInfo.Options);
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectCodeGenConverter.cs:39]: nit: do we need the bang?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs [hunk 3]
```diff
@@ -4,13 +4,14 @@
...
 using System.Runtime.CompilerServices;
+using System.Text.Json.Serialization.Metadata;
 
 namespace System.Text.Json.Serialization.Converters
 {
     /// <summary>
     /// Default base class implementation of <cref>JsonObjectConverter{T}</cref>.
     /// </summary>
-    internal class ObjectDefaultConverter<T> : JsonObjectConverter<T> where T : notnull
+    internal class ObjectDefaultConverter<T> : JsonObjectConverter<T>
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs:14]: why the change to `notnull`?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 4]
```diff
@@ -68,6 +70,8 @@ internal override sealed JsonParameterInfo CreateJsonParameterInfo()
         /// </summary>
         internal bool CanBeNull { get; }
 
+        internal readonly EqualityComparer<T> _defaultComparer = EqualityComparer<T>.Default;
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:73]: is this used?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonValueInfoOfT.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        /// <summary>
+        /// todo
+        /// </summary>
+        public void CompleteInitialization()
+        {
+            _isInitialized = true;
+
+            //todo: should we not add?
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonValueInfoOfT.cs:33]: What is the benefit of adding given all the generated TypeInfos have references to each other via their `JsonPropertyInfo`s?

File: src/libraries/System.Text.Json/tests/Serialization/CodeGenTests/SlowPathTests.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                    jsonTypeInfo: typeInfo);
+
+                Assert.Equal(Count, newObjs.Count);
+                for (int i = 0; i < objs.Count; i++)
+                {
+                    IndexViewModelPocoTests.Verify(objs[i], newObjs[i]);
+                }
+            }
+
+            // todo: add SerializeAsync method
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/CodeGenTests/SlowPathTests.cs:56]: 👀

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs:14]: I'm not sure why this was originally added; affects new usage of generics added elsewhere. `T` can be null...

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 4 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:73]: Yes in JsonPropertyInfo `WriteValue` and `Write`. I believe you refactored this earlier in master; not sure if this branch is up-to-date with master to have that change yet.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonValueInfoOfT.cs (hunk 5 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonValueInfoOfT.cs:33]: It allows direct usage of the type in the existing (De)Serialize methods without having to fall back to reflection\IL Emit, and also allows `options.GetConverter(type)` to return the type without reflection\IL Emit.

File: src/libraries/System.Text.Json/tests/Serialization/CodeGenTests/SlowPathTests.cs (hunk 6 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/CodeGenTests/SlowPathTests.cs:56]: As a proof-of-concept, just the DeserializeAsync was added (to ensure callers can still use that, even though the generated (de)serialize methods won't get called. The SerializeAsync will be added next pass.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs]: Not needed. Thanks

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectCodeGenConverter.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectCodeGenConverter.cs]: No. Thanks

**Comment by steveharter** [comment]
Test failure is not related: System.Net.Http.Functional.Tests

**Pull request merged**